File: hdf-cli/pkg/diff/validate/validate_test.go
package validate

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	engine "github.com/mitre/hdf-cli/pkg/diff/engine"
	"github.com/mitre/hdf-cli/pkg/diff/normalize"
	types "github.com/mitre/hdf-cli/pkg/diff/types"
	hdf "github.com/mitre/hdf-cli/pkg/hdf"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// Fixture helpers (same pattern as engine/differential_test.go).

// fixturesDir returns the absolute path to the shared test fixtures directory.
func fixturesDir(t *testing.T) string {
	t.Helper()
	_, filename, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	// From hdf-cli/pkg/diff/validate/ up to hdf-libs root, then into hdf-diff/test/fixtures.
	dir := filepath.Join(filepath.Dir(filename), "..", "..", "..", "..", "hdf-diff", "test", "fixtures")
	absDir, err := filepath.Abs(dir)
	if err != nil {
		t.Fatalf("filepath.Abs failed: %v", err)
	}
	return absDir
}

// loadV2Fixture loads a v2-format HDF JSON fixture into hdf.HdfResults.
func loadV2Fixture(t *testing.T, name string) hdf.HdfResults {
	t.Helper()
	path := filepath.Join(fixturesDir(t), name)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Skipf("fixture %s not found: %v (run from monorepo root)", name, err)
	}
	result, err := hdf.UnmarshalHdfResults(data)
	if err != nil {
		t.Fatalf("failed to parse v2 fixture %s: %v", name, err)
	}
	return result
}

// loadV1Fixture loads a v1-format InSpec exec-json fixture, normalizes it to v2,
// and returns hdf.HdfResults.
func loadV1Fixture(t *testing.T, name string) hdf.HdfResults {
	t.Helper()
	path := filepath.Join(fixturesDir(t), name)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Skipf("fixture %s not found: %v (run from monorepo root)", name, err)
	}
	var raw map[string]any
	if err := json.Unmarshal(data, &raw); err != nil {
		t.Fatalf("failed to parse JSON for %s: %v", name, err)
	}
	if !normalize.IsV1Format(raw) {
		t.Fatalf("expected %s to be v1 format", name)
	}
	result, _, err := normalize.ToV2(data)
	if err != nil {
		t.Fatalf("failed to normalize v1 fixture %s: %v", name, err)
	}
	return result
}

// runDiff runs the diff engine on two HDF results documents with default options.
func runDiff(t *testing.T, oldDoc, newDoc hdf.HdfResults) types.HdfComparison {
	t.Helper()
	comp, err := engine.DiffHdf(oldDoc, []hdf.HdfResults{newDoc}, engine.Options{})
	require.NoError(t, err, "DiffHdf should not return an error")
	return comp
}

// comparisonToAny marshals an HdfComparison to JSON and back to any (map),
// which is the format the validator expects.
//
// Note: ValidateComparison now calls NormalizeForSchema internally,
// so this helper no longer needs to normalize manually. It just
// marshals the Go struct to a generic map[string]any for the validator.
func comparisonToAny(t *testing.T, comp types.HdfComparison) any {
	t.Helper()
	data, err := json.Marshal(comp)
	require.NoError(t, err, "json.Marshal should not fail")
	var doc any
	err = json.Unmarshal(data, &doc)
	require.NoError(t, err, "json.Unmarshal should not fail")
	return doc
}

// Test 1: Valid output from DiffHdf validates successfully.
func TestValidateComparison_DiffOutput(t *testing.T) {
	scanBefore := loadV2Fixture(t, "scan-before.json")
	scanAfter := loadV2Fixture(t, "scan-after.json")
	comp := runDiff(t, scanBefore, scanAfter)
	doc := comparisonToAny(t, comp)

	result := ValidateComparison(doc)
	if !result.Valid {
		t.Logf("Validation errors: %v", result.Errors)
	}
	assert.True(t, result.Valid, "DiffHdf output should validate against schema")
	assert.Empty(t, result.Errors, "No errors expected for valid output")
}

// Test 2: Valid identical comparison validates.
func TestValidateComparison_Identical(t *testing.T) {
	scanBefore := loadV2Fixture(t, "scan-before.json")
	comp := runDiff(t, scanBefore, scanBefore)
	doc := comparisonToAny(t, comp)

	result := ValidateComparison(doc)
	if !result.Valid {
		t.Logf("Validation errors: %v", result.Errors)
	}
	assert.True(t, result.Valid, "identical comparison should validate")
}

// Test 3: Valid fleet comparison validates.
func TestValidateComparison_Fleet(t *testing.T) {
	scanBefore := loadV2Fixture(t, "scan-before.json")
	scanAfter := loadV2Fixture(t, "scan-after.json")

	comp, err := engine.DiffHdf(scanBefore, []hdf.HdfResults{scanAfter, scanBefore}, engine.Options{
		ComparisonMode: types.ModeFleet,
	})
	require.NoError(t, err)
	doc := comparisonToAny(t, comp)

	result := ValidateComparison(doc)
	if !result.Valid {
		t.Logf("Validation errors: %v", result.Errors)
	}
	assert.True(t, result.Valid, "fleet mode comparison should validate")
}

// Test 4: Invalid document rejected (missing required fields).
func TestValidateComparison_InvalidDocument(t *testing.T) {
	doc := map[string]any{"foo": "bar"}
	result := ValidateComparison(doc)

	assert.False(t, result.Valid, "invalid document should be rejected")
	assert.NotEmpty(t, result.Errors, "errors should be present")
	assert.Greater(t, len(result.Errors), 0, "at least one error expected")
}

// Test 5: Missing formatVersion rejected.
func TestValidateComparison_MissingFormatVersion(t *testing.T) {
	doc := map[string]any{"formatVersion": "1.0.0"}
	result := ValidateComparison(doc)

	assert.False(t, result.Valid, "document with only formatVersion should be rejected")
	assert.NotEmpty(t, result.Errors, "errors should be present for missing required fields")
}

// Test 6: Invalid comparisonMode rejected.
func TestValidateComparison_InvalidComparisonMode(t *testing.T) {
	scanBefore := loadV2Fixture(t, "scan-before.json")
	scanAfter := loadV2Fixture(t, "scan-after.json")
	comp := runDiff(t, scanBefore, scanAfter)
	doc := comparisonToAny(t, comp)

	// Modify to invalid comparison mode.
	docMap := doc.(map[string]any)
	docMap["formatVersion"] = "2.0.0"
	result := ValidateComparison(docMap)

	assert.False(t, result.Valid, "document with invalid formatVersion should be rejected")
}

// Test 7: Nginx v1 comparison validates.
func TestValidateComparison_NginxV1(t *testing.T) {
	nginxFailing := loadV1Fixture(t, "nginx-failing.json")
	nginxClean := loadV1Fixture(t, "nginx-clean.json")
	comp := runDiff(t, nginxFailing, nginxClean)
	doc := comparisonToAny(t, comp)

	result := ValidateComparison(doc)
	if !result.Valid {
		errCount := len(result.Errors)
		if errCount > 5 {
			errCount = 5
		}
		t.Logf("Validation errors (first %d): %v", errCount, result.Errors[:errCount])
	}
	assert.True(t, result.Valid, "nginx (failing -> clean) comparison should validate")
}

// Test 8: Ubuntu comparison validates.
func TestValidateComparison_UbuntuV1(t *testing.T) {
	vanilla := loadV1Fixture(t, "ubuntu-22-vanilla.json")
	hardened := loadV1Fixture(t, "ubuntu-22-hardened.json")
	comp := runDiff(t, vanilla, hardened)
	doc := comparisonToAny(t, comp)

	result := ValidateComparison(doc)
	if !result.Valid {
		errCount := len(result.Errors)
		if errCount > 5 {
			errCount = 5
		}
		t.Logf("Validation errors (first %d): %v", errCount, result.Errors[:errCount])
	}
	assert.True(t, result.Valid, "ubuntu (vanilla -> hardened) comparison should validate")
}

// Test 9: Empty object rejected.
func TestValidateComparison_EmptyObject(t *testing.T) {
	doc := map[string]any{}
	result := ValidateComparison(doc)

	assert.False(t, result.Valid, "empty object should be rejected")
	assert.NotEmpty(t, result.Errors, "errors should be present for empty object")
}

// Test 10: Error messages are descriptive strings.
func TestValidateComparison_DescriptiveErrors(t *testing.T) {
	doc := map[string]any{"foo": "bar"}
	result := ValidateComparison(doc)

	require.False(t, result.Valid)
	require.NotEmpty(t, result.Errors)

	for _, errMsg := range result.Errors {
		assert.IsType(t, "", errMsg, "each error should be a string")
		assert.NotEmpty(t, errMsg, "error messages should not be empty")
		// Error messages should contain meaningful information.
		assert.Greater(t, len(errMsg), 5, "error messages should be descriptive (more than 5 chars)")
	}
}

File: hdf-diff/test/validate.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { diffHdf } from '../src/diff.js';
import { validateComparison } from '../src/validate.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

function loadFixture(name: string): Record<string, unknown> {
  const filePath = resolve(__dirname, 'fixtures', name);
  return JSON.parse(readFileSync(filePath, 'utf-8')) as Record<string, unknown>;
}

describe('validateComparison', () => {
  let scanBefore: Record<string, unknown>;
  let scanAfter: Record<string, unknown>;

  beforeAll(() => {
    scanBefore = loadFixture('scan-before.json');
    scanAfter = loadFixture('scan-after.json');
  });

  // ── Valid output validates ────────────────────────────────────────
  describe('valid output validates', () => {
    it('should validate output from diffHdf(scanBefore, scanAfter)', () => {
      const comparison = diffHdf(scanBefore, scanAfter);
      const result = validateComparison(comparison);
      if (!result.valid) {
        // Log errors for debugging
        console.error('Validation errors:', result.errors);
      }
      expect(result.valid).toBe(true);
      expect(result.errors).toBeUndefined();
    });

    it('should validate output from diffHdf(scanBefore, scanBefore) — identical', () => {
      const comparison = diffHdf(scanBefore, scanBefore);
      const result = validateComparison(comparison);
      if (!result.valid) {
        console.error('Validation errors:', result.errors);
      }
      expect(result.valid).toBe(true);
    });
  });

  // ── Fleet mode output validates ──────────────────────────────────
  describe('fleet mode output validates', () => {
    it('should validate fleet mode comparison output', () => {
      const comparison = diffHdf(scanBefore, [scanAfter, scanBefore], {
        comparisonMode: 'fleet',
      });
      const result = validateComparison(comparison);
      if (!result.valid) {
        console.error('Validation errors:', result.errors);
      }
      expect(result.valid).toBe(true);
    });
  });

  // ── Invalid documents rejected ───────────────────────────────────
  describe('invalid document rejection', () => {
    it('should reject a completely invalid document', () => {
      const result = validateComparison({ foo: 'bar' });
      expect(result.valid).toBe(false);
      expect(result.errors).toBeDefined();
      expect(result.errors!.length).toBeGreaterThan(0);
    });

    it('should reject a document missing required fields', () => {
      const result = validateComparison({ formatVersion: '1.0.0' });
      expect(result.valid).toBe(false);
      expect(result.errors).toBeDefined();
      expect(result.errors!.length).toBeGreaterThan(0);
    });

    it('should reject a document with wrong formatVersion', () => {
      const comparison = diffHdf(scanBefore, scanAfter) as Record<string, unknown>;
      const modified = { ...comparison, formatVersion: '2.0.0' };
      const result = validateComparison(modified);
      expect(result.valid).toBe(false);
    });
  });

  // ── Nginx v1 comparison validates ────────────────────────────────
  describe('nginx v1 fixture comparison validates', () => {
    it('should validate nginx (failing -> clean) comparison', () => {
      const nginxFailing = loadFixture('nginx-failing.json');
      const nginxClean = loadFixture('nginx-clean.json');
      const comparison = diffHdf(nginxFailing, nginxClean);
      const result = validateComparison(comparison);
      if (!result.valid) {
        console.error('Validation errors (first 5):', result.errors?.slice(0, 5));
      }
      expect(result.valid).toBe(true);
    });
  });

  // ── Ubuntu comparison validates ──────────────────────────────────
  describe('ubuntu fixture comparison validates', () => {
    it('should validate ubuntu (vanilla -> hardened) comparison', () => {
      const vanilla = loadFixture('ubuntu-22-vanilla.json');
      const hardened = loadFixture('ubuntu-22-hardened.json');
      const comparison = diffHdf(vanilla, hardened);
      const result = validateComparison(comparison);
      if (!result.valid) {
        console.error('Validation errors (first 5):', result.errors?.slice(0, 5));
      }
      expect(result.valid).toBe(true);
    });
  });

  // ── diffHdf with validateOutput option ───────────────────────────
  describe('diffHdf with validateOutput option', () => {
    it('should not throw when validateOutput is true and output is valid', () => {
      expect(() =>
        diffHdf(scanBefore, scanAfter, { validateOutput: true }),
      ).not.toThrow();
    });

    it('should return a valid HDFComparison when validateOutput is true', () => {
      const comparison = diffHdf(scanBefore, scanAfter, { validateOutput: true });
      expect(comparison.formatVersion).toBe('1.0.0');
      expect(comparison.requirementDiffs.length).toBeGreaterThan(0);
    });
  });
});
